modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtCacheAdapter.java

```diff
@@ -112,76 +105,29 @@
     /** */
     private static final long serialVersionUID = 0L;
 
-    /** Force key futures. */
-    private final ConcurrentMap<IgniteUuid, GridDhtForceKeysFuture<?, ?>> forceKeyFuts = newMap();
-
-    /** */
-    private volatile boolean stopping;
-
-    /** Discovery listener. */
-    private final GridLocalEventListener discoLsnr = new GridLocalEventListener() {
-        @Override public void onEvent(Event evt) {
-            DiscoveryEvent e = (DiscoveryEvent)evt;
-
-            ClusterNode loc = ctx.localNode();
-
-            assert e.type() == EVT_NODE_LEFT || e.type() == EVT_NODE_FAILED : e;
-
-            final ClusterNode n = e.eventNode();
-
-            assert !loc.id().equals(n.id());
-
-            for (GridDhtForceKeysFuture<?, ?> f : forceKeyFuts.values())
-                f.onDiscoveryEvent(e);
-        }
-    };
-
     /**
-     * Empty constructor required for {@link Externalizable}.
+     * TODO GG-30033 remove this method.
      */
-    protected GridDhtCacheAdapter() {
-        // No-op.
-    }
+    protected void assertOwning(GridCacheContext cctx, Collection<KeyCacheObject> keys, AffinityTopologyVersion topVer) {
+        HashSet<Integer> parts = new HashSet<>(keys.size(), 1.f);
 
-    /**
-     * Adds future to future map.
-     *
-     * @param fut Future to add.
-     * @return {@code False} if node cache is stopping and future was completed with error.
-     */
-    public boolean addFuture(GridDhtForceKeysFuture<?, ?> fut) {
-        forceKeyFuts.put(fut.futureId(), fut);
+        for (KeyCacheObject key : keys) {
+            int part = cctx.affinity().partition(key);
 
-        if (stopping) {
-            fut.onDone(stopError());
+            if (parts.add(part)) {
+                GridDhtLocalPartition dPart = cctx.topology().localPartition(part, topVer, false);
 
-            return false;
+                assert dPart != null && dPart.state() == GridDhtPartitionState.OWNING :
+                    "Invalid partition state [part=" + dPart + ", key=" + key + ", topVer=" + topVer + ']';
+            }
         }
-
-        return true;
-    }
-
-    /**
-     * Removes future from future map.
-     *
-     * @param fut Future to remove.
-     */
-    public void removeFuture(GridDhtForceKeysFuture<?, ?> fut) {
-        forceKeyFuts.remove(fut.futureId(), fut);
     }
 
     /**
-     * @param node Node.
-     * @param msg Message.
+     * Empty constructor required for {@link Externalizable}.
      */
-    protected final void processForceKeyResponse(ClusterNode node, GridDhtForceKeysResponse msg) {
-        GridDhtForceKeysFuture<?, ?> f = forceKeyFuts.get(msg.futureId());
-
-        if (f != null)
-            f.onResult(msg);
-        else if (log.isDebugEnabled())
-            log.debug("Receive force key response for unknown future (is it duplicate?) [nodeId=" + node.id() +
-                ", res=" + msg + ']');
+    protected GridDhtCacheAdapter() {
```

Comment:
Let's create a ticket to unimplement `Externalizable`, from my point of view it's impossible to serialize cache adapter and I can't imagine a case when we need it.

Reply:
https://issues.apache.org/jira/browse/IGNITE-13241

Reply:
We only need proxies to be externalizable, no need to have the cache adapter to use this interface. I put this into the ticket.

Reply:
Ok, we can do this safely for Ignite but not for GG, or it will break compatibility.